src/map.rs

```diff
@@ -1020,15 +1020,13 @@ where
                         // the new bin will also be a tree bin. if both the high
                         // bin and the low bin are non-empty, we have to
                         // allocate a new TreeBin.
-                        Shared::boxed(
-                            BinEntry::Tree(TreeBin::new(
-                                // safety: we have just created `low` and its `next`
-                                // nodes and have never shared them
-                                unsafe { low.into_box() },
-                                guard,
-                            )),
-                            &self.collector,
-                        )
+                        // safety: we have just created `low` and its `next`
+                        // nodes and have never shared them
+                        let low_bin = unsafe {
+                            BinEntry::Tree(TreeBin::new(Box::into_raw(low.into_box()), guard))
```

Comment:
Based on my read of the discussion, I think we shouldn't `Box::into_raw(low.into_box())` here (and elsewhere) — we shouldn't be re-constructing a `Box` at all. Instead, we should grab the raw pointer directly with `low.0.into_inner()` so that we preserve the right provenance. In fact, we should probably remove `fn into_box` entirely.